File: listings/src/main/java/chapter05/Listing0518.java
package chapter05;

public class Listing0518 {
  public static void main(String[] args) {
    Cache0518.cache("supercar", new Car());      // Static member is callable directly on class
    Cache0518.INSTANCE.cache("car", new Car());  // Bad practice

    Car0518.produceCar();                        // Static
    Car0518.Factory.produceCar();                // Also possible
    (new Car0518()).produceCar();                // Bad practice

  }
}

File: listings/src/main/kotlin/chapter05/Listing0518.kt
package chapter05

// Must rename declarations to avoid name clashes in project
object Cache0518 {
  @JvmStatic fun cache(key: String, obj: Any) { }  // Becomes a static member
}

class Car0518 {
  companion object Factory {
    @JvmStatic fun produceCar() { }                  // Now becomes static as well
  }
}
